### PR TITLE
Make flockdrones hibernate if they have nothing to do

### PR DESCRIPTION
**Pull request by amylizzle**
## About the PR 
If a drone has spent the last few cycles of it's AI wandering, it will instead go into hibernation. This causes it to a) stop eating tons of CPU doing JPS wandering, and b) provide 5 bonus compute to the flock.

Drones will wake from hibernation on being given specific orders, or on detecting an enemy nearby (this is done on the `Life()` loop, so every few seconds or so).

## Why's this needed? 
Flock basically stops moving towards the end of the round because of all the drones doing pathing. This should substantially reduce that.

## Changelog

```changelog
(u)Amylizzle
(+)Flockdrones with nothing to do will now hibernate, providing 5 extra compute. They will wake from hibernation when given orders or an enemy is detected.
```

### CODE REVIEW

File: code/mob/living/critter/ai/flock/flocktasks.dm [hunk 1]
```diff
@@ -94,6 +94,11 @@ stare
 	src.subtasks = list() //get rid of the move and replace it with flockmove
 	add_task(holder.get_instance(/datum/aiTask/succeedable/move/flock, list(holder)))
 
+/datum/aiTask/sequence/goalbased/flock/switched_to()
+	. = ..()
+	var/mob/living/critter/flock/drone/D = holder.owner
+	if(istype(D))
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/ai/flock/flocktasks.dm:100]: This PR contains three different ways of casting `holder.owner` to `/mob/living/critter/flock/drone`, please gods pick one

File: code/mob/living/critter/flock/flockdrone.dm [hunk 2]
```diff
@@ -278,6 +283,39 @@
...
 		src.is_npc = TRUE
 
+/mob/living/critter/flock/drone/proc/pause_ai()
+	if(src.controller) //can't pause_ai when controlled, this shouldn't ever happen
+		return
+	src.ai.stop_move()
+	src.ai_paused = TRUE
+	src.icon_state = "drone-dormant"
+	src.remove_simple_light("drone_light")
+	flock_speak(src, "No tasks in queue. Allocating higher functions to resource generation.", src.flock)
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/flock/flockdrone.dm:293]:
```suggestion
	flock_speak(src, "No tasks in queue. Allocating higher functions to compute generation.", src.flock)
```

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by FlameArrow57** [comment]
Are there any more AI optimizations that can be done? I just kind of hate that we have to resort to making drones hibernate if that's how intensive the AI is going to be

**Comment by amylizzle** [comment]
There's probably a few tweaks to be made, but I already did a pretty thorough pass. The problem is just that pathfinding is expensive, and doing it a lot slows things down. Short of caching the pathfinding results (won't work, things changing too much), there's not a lot that can be done.

Plus, you do always end up with like 50 drones just wandering at the end of the round because they've eaten everything and converted everything. No point wasting all those cycles.

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by amylizzle** [comment]
!merge_upstream

**Comment by FlameArrow57** [comment]
Seems critters already have an is_hibernating variable, not sure if that could be reused if wanted

In Discord it was mentioned Life() ticks like every 12 seconds or something on main, feel like it could also be good to add it to TakeDamage() too, or maybe was_harmed, might be good to integrate with the COMSIG_MOB_GRABBED signal for waking too otherwise think a hibernating drone probably shouldn't shock a person

Also need to add waking to drone/Login()

Need to add to handle_flock_attack() proc so that hibernating drones don't report attacks

Need to add to check_health() proc to not update icon when hibernating

Also need to add to flock/Life() to not use extinguisher while ai is paused, could also maybe wake drone up if on fire

If drones can eat items while wandering, probably should either prevent hibernating while eating items, or stop items from being eaten while hibernating

Probably should make it so that drones that are dormantized don't report going dormant if hibernating

Also not sure if drones that are hibernating should report death upon dying

File: _std/macros/flock.dm [hunk 3]
```diff
@@ -33,3 +33,9 @@
 // replication defs
 #define FLOCK_MIN_DESIRED_POP 10
 #define FLOCK_ADDITIONAL_RESOURCE_RESERVATION_PER_DRONE 7.5
+
+// drone compute configs
+#define FLOCK_DRONE_COMPUTE 10
+#define FLOCK_DRONE_COMPUTE_HIBERNATE 15
+// how many times a drone must wander in a row before it hibernates
+#define FLOCK_DRONE_WANDER_PAUSE_COUNT 5
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply _std/macros/flock.dm:41]: I put them as defines so they'd be easily configurable for future balance changes without needing to dig into the code. I've found it pretty useful when debugging and tuning this.

File: code/mob/living/critter/flock/flockdrone.dm [hunk 4]
```diff
@@ -278,6 +284,38 @@
...
+/mob/living/critter/flock/drone/proc/pause_ai()
+	if(src.controller) //can't pause_ai when controlled, this shouldn't ever happen
+		return
+	src.ai.stop_move()
+	src.ai_paused = TRUE
+	src.icon_state = "drone-dormant"
+	src.remove_simple_light("drone_light")
+	flock_speak(src, "No tasks in queue. Allocating higher functions to compute generation.", src.flock)
+	src.is_npc = FALSE
+	src.compute = FLOCK_DRONE_COMPUTE_HIBERNATE
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply code/mob/living/critter/flock/flockdrone.dm:296]: That PR isn't merged yet?

File: code/mob/living/critter/flock/flockdrone.dm [hunk 5]
```diff
@@ -278,6 +284,38 @@
...
+
+/mob/living/critter/flock/drone/proc/wake_from_ai_pause()
+	if(!src.ai_paused)
+		return
+	src.compute = FLOCK_DRONE_COMPUTE
+	src.ai_paused = FALSE
+	src.canmove = TRUE
+	src.anchored = FALSE
+	src.wander_count = 0
+	src.damaged = -1 //force icon refresh
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply code/mob/living/critter/flock/flockdrone.dm:308]: It does, setting damaged to -1 triggers the icon update back to the appropriate level for the drone's health
```
	switch(percent_damage)
		if(75 to 100)
			if(damaged == 0) return
			damaged = 0
			if(!dormant)
				src.icon_state = "drone"
```

File: code/mob/living/critter/flock/flockdrone.dm [hunk 6]
```diff
@@ -278,6 +284,38 @@
...
+	src.check_health() // handles updating the icon to something more appropriate
+	src.visible_message("<span class='notice'><b>[src]</b> begins to glow and hover.</span>")
+	src.set_a_intent(INTENT_HELP)
+	src.add_simple_light("drone_light", rgb2num(glow_color))
+	if(src.client)
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply code/mob/living/critter/flock/flockdrone.dm:312]: agreed

File: code/mob/living/critter/flock/flockdrone.dm [hunk 7]
```diff
@@ -429,7 +467,20 @@
 		return
 	if (src.dormant)
 		return
-
+	if(src.ai_paused)
+		//wake up if there are enemies in view
+		if(src.flock) //if we have a flock, use the enemies list, otherwise just use non-flock mobs in view
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply code/mob/living/critter/flock/flockdrone.dm:471]: Flock could die, or the drone could have been spawned without a flock (azone or something)

File: code/mob/living/critter/flock/flockdrone.dm [hunk 8]
```diff
@@ -429,7 +467,20 @@
...
+				if(enemy in viewers(src))
+					src.wake_from_ai_pause()
+					break
+		else
+			for(var/mob/living/carbon/human/enemy in viewers(src))
+				if(!isdead(enemy))
+					src.wake_from_ai_pause()
+					break
+	else if(src.wander_count > FLOCK_DRONE_WANDER_PAUSE_COUNT)
+		src.pause_ai()
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply code/mob/living/critter/flock/flockdrone.dm:482]: I mean, if you've wandered 5 times in a row already, it's not like there's a lot changing there. I think it's fine to check this sparingly.

**Comment by amylizzle** [comment]
> Seems critters already have an is_hibernating variable, not sure if that could be reused if wanted

Yeah, I looked into it when I started this one. It does too much other stuff and is used for azone area stuff. Too different to be useful unfortunately.

 
> In Discord it was mentioned Life() ticks like every 12 seconds or something on main, feel like it could also be good to add it to TakeDamage() too, or maybe was_harmed, might be good to integrate with the COMSIG_MOB_GRABBED signal for waking too otherwise think a hibernating drone probably shouldn't shock a person

Life() ticks commented as above, I think it'll be fine. Adding it to TakeDamage() and the like is a good idea, will do.
 
> Also need to add waking to drone/Login()

ditto 

> Need to add to handle_flock_attack() proc so that hibernating drones don't report attacks

I think it makes sense for them to report attacks, since they are _aware_ of enemies - that's how they wake up. Reporting an attack could wake them up too.

> Need to add to check_health() proc to not update icon when hibernating

smort.
 
> Also need to add to flock/Life() to not use extinguisher while ai is paused, could also maybe wake drone up if on fire

Waking up because you're on fire seems reasonable.

> If drones can eat items while wandering, probably should either prevent hibernating while eating items, or stop items from being eaten while hibernating

I'm fine with passive absorbtion, but this should never happen because drones will have to have been wandering for 5 cycles first which means they haven't done any harvesting recently.

> Probably should make it so that drones that are dormantized don't report going dormant if hibernating

will do

> Also not sure if drones that are hibernating should report death upon dying

Yeah, they ought to. Hibernation is just a less active state, there's still some awareness there.

**Comment by flrsh** [comment]
!merge_upstream

**Review by flrsh — APPROVED** [review]
Briefly reviewed and looks good to me!

**Pull request merged**